File: AllInOne/src/DyanmicProgramming/MaximumSubarray.java
/**
 * 
 * 53. Maximum Subarray
 *
 * @author jingjie jiang
 * @history Apr 24, 2022
 *
 */
class Solution {
    public int maxSubArray(int[] nums) {
        
        assert nums != null && nums.length >= 1 && nums.length <= Math.pow(10, 5);
        
        int maxSum = nums[0], preSum = nums[0];
        
        for (int idx = 1; idx < nums.length; idx ++) {
            
            int curSum = preSum + nums[idx];
            preSum = Math.max(curSum, nums[idx]);
            maxSum = Math.max(preSum, maxSum);
        }
        
        return maxSum;
    }
}

File: AllInOne/src/DyanmicProgramming/MaximumSubarray.js
/**
 * 
 * 53. Maximum Subarray
 *
 */
/**
 * @param {number[]} nums
 * @return {number}
 */
 var maxSubArray = function(nums) {
    if (!nums || nums.length === 0) return 0;

    let max = Number.MAX_SAFE_INTEGER;
    let curSum = 0;

    for (let idx = 0; idx < nums.length; idx ++) {
      curSum += nums[idx];
      max = Math.max(curSum, max);
      curSum = curSum < 0 ? 0 : curSum;
    }

    return max;
};